Rename category type field in ElectricalComponent

tiyash-basu-frequenz opened the pull request:
This commit renames the folowing:
- `ComponentCatgoryMetadataVariant` to `ElectricalComponentCategorySpecificInfo`
- `ElectricalComponent.category_type` to `ElectricalComponent.category_specific_info`

These renames make the entity names better reflect their purpose.

llucax commented:
You sneaked the unification of metric and metric sample with sensors, which seems like a major changeworth of mentioning in the PR (arguably more relevant than a simple rename).

llucax commented on proto/frequenz/api/common/v1/metrics/metrics.proto:
```diff
@@ -22,7 +22,7 @@ message SimpleMetricValue {
 
 // Encapsulates derived statistical summaries of a single metric.
-//
+c//
```
Whoops!

tiyash-basu-frequenz replied:
This commit was added by mistake and has now been removed.

llucax commented on proto/frequenz/api/common/v1/microgrid/sensors/sensors.proto:
```diff
@@ -166,7 +166,7 @@ message SensorDiagnostic {
   string message = 3;
 }
-
+bopids
```
Whooopy whoops!

tiyash-basu-frequenz replied:
This commit was added by mistake and has now been removed.

llucax commented on proto/frequenz/api/common/v1/microgrid/sensors/sensors.proto:
Shouldn't `SensorMetric` be completely removed from here too?

tiyash-basu-frequenz replied:
This commit was added by mistake and has now been removed.

tiyash-basu-frequenz commented:
> You sneaked the unification of metric and metric sample with sensors, which seems like a major changeworth of mentioning in the PR (arguably more relevant than a simple rename).

No that is a mistake. The commits should not be there. Will remove those.

tiyash-basu-frequenz commented on RELEASE_NOTES.md:
```diff
@@ -22,6 +22,8 @@
 - Added new message `CommunicationComponentDiagnostic` to represent warnings and errors in microgrid communication components.
 - Added new message `CommunicationComponentStateSnapshot` to represent the state of communication components.
 - Added new message definitions for streaming events (Deleted, Created, Updated)
+- Renamed `ComponentCatgoryMetadataVariant` to `ElectricalComponentCategorySpecificInfo`.
```
No, the text I wrote is correct w.r.t. the overall changes in this milestone.

tiyash-basu-frequenz commented on RELEASE_NOTES.md:
```diff
@@ -22,6 +22,8 @@
 - Added new message `CommunicationComponentDiagnostic` to represent warnings and errors in microgrid communication components.
 - Added new message `CommunicationComponentStateSnapshot` to represent the state of communication components.
 - Added new message definitions for streaming events (Deleted, Created, Updated)
+- Renamed `ComponentCategoryMetadataVariant` to `ElectricalComponentCategorySpecificInfo`.
```
No, the text I wrote is correct w.r.t. the overall changes in this milestone.

llucax commented on proto/frequenz/api/common/v1/microgrid/electrical_components/electrical_components.proto:
```diff
@@ -349,8 +349,8 @@ message MetricConfigBounds {
-message ElectricalComponentCategoryMetadataVariant {
-  oneof metadata {
+message ElectricalComponentCategorySpecificInfo {
+  oneof info {
```
Super low importance, but while we are at this, and while researching some other `oneof`, I got the impression `kind` is pretty standard[^1] mainly when a message only contains one `oneof` and the name becomes redundant with the message name, so maybe we can use that here too?

I also asked Gemini, and they recommend `component` or `details`.

<details>
<summary>Gemini response</summary>

Okay, this is a great, concrete example to discuss\! Let's look at your `ElectricalComponentCategorySpecificInfo` message:

```protobuf
// Metadata specific to a microgrid component.
message ElectricalComponentCategorySpecificInfo {
  oneof info { // <-- This is the oneof in question
    frequenz.api.common.v1.microgrid.electrical_components.Battery battery = 1;
    frequenz.api.common.v1.microgrid.electrical_components.EvCharger ev_charger = 2;
    frequenz.api.common.v1.microgrid.electrical_components.Fuse fuse = 3;
    frequenz.api.common.v1.microgrid.electrical_components.GridConnectionPoint grid = 4;
    frequenz.api.common.v1.microgrid.electrical_components.Inverter inverter = 5;
    frequenz.api.common.v1.microgrid.electrical_components.VoltageTransformer voltage_transformer = 6;
  }
}
```

The containing message is `ElectricalComponentCategorySpecificInfo`. This tells us that the message holds information specific to a component category. The `oneof` then provides the actual data for *one* of those specific component types.

Let's evaluate the current name `info` and some alternatives:

1.  **`info` (current):**

      * Access path: `...category_specific_info.info.battery`
      * Pros: Short.
      * Cons: It's a bit redundant. "Electrical Component Category Specific Info's info" sounds like "the information of the specific information." While not terrible, it could be more distinct.

2.  **`kind`:**

      * Access path: `...category_specific_info.kind.battery`
      * Pros: This is a standard and widely understood convention for `oneof` fields that act as discriminated unions, indicating the "kind" of data present. It signifies that the specific info can be *of the kind* "battery," "ev\_charger," etc.
      * Cons: It's a bit abstract.

3.  **`type`:**

      * Access path: `...category_specific_info.type.battery`
      * Pros: Similar to `kind`, very common. It clearly indicates that this field determines the specific *type* of the component information.
      * Cons: Also somewhat abstract.

4.  **`component`:**

      * Access path: `...category_specific_info.component.battery`
      * Pros: This is quite descriptive in your domain. The `oneof` is literally selecting *which specific component's detailed message* is being provided. The fields within the `oneof` are `battery`, `ev_charger` – these are types of components. So, saying `category_specific_info.component` clearly points to "the actual component data within this specific info structure."
      * Cons: The word "component" is already part of the message name and the broader context (`ElectricalComponent`). However, this isn't necessarily a strong con, as it reinforces the domain concept.

5.  **`details`:**

      * Access path: `...category_specific_info.details.battery`
      * Pros: The `oneof` holds the *details* for the specific category. This avoids the "info.info" redundancy and is quite intuitive. "Category-specific info's details" makes good sense.
      * Cons: A bit more generic than `component`.

6.  **`data`:**

      * Access path: `...category_specific_info.data.battery`
      * Pros: Similar to `details`. The `oneof` contains the actual *data* payload for the specific category. Clear and common for payload-carrying fields.
      * Cons: Also somewhat generic.

**Recommendation:**

Given your specific structure and domain:

  * **Option A: `component`**
    This is a strong contender. The `oneof` is selecting from different *component* types. `category_specific_info.component.battery` reads very naturally as "the battery component data within the category-specific information." It's domain-relevant and clearly indicates the nature of the choice.

  * **Option B: `details` or `data`**
    These are also good choices. `category_specific_info.details.battery` emphasizes that this `oneof` provides the specific, detailed information for the category. It elegantly avoids the redundancy of the original `info` name.

  * **Option C: `kind` or `type`**
    These are safe, standard choices that any Protobuf developer would immediately understand. They emphasize the polymorphic nature of the `ElectricalComponentCategorySpecificInfo` message.

**My top recommendation would be `component`.**

```protobuf
// Metadata specific to a microgrid component.
message ElectricalComponentCategorySpecificInfo {
  oneof component { // Renamed from 'info' to 'component'
    frequenz.api.common.v1.microgrid.electrical_components.Battery battery = 1;
    frequenz.api.common.v1.microgrid.electrical_components.EvCharger ev_charger = 2;
    // ... and so on
  }
}
```

This feels the most direct and semantically aligned with what the `oneof` represents: the specific component type whose data is included.

**Second choice would be `details`:**

```protobuf
// Metadata specific to a microgrid component.
message ElectricalComponentCategorySpecificInfo {
  oneof details { // Renamed from 'info' to 'details'
    frequenz.api.common.v1.microgrid.electrical_components.Battery battery = 1;
    frequenz.api.common.v1.microgrid.electrical_components.EvCharger ev_charger = 2;
    // ... and so on
  }
}
```

This is also very clear and avoids any redundancy.

Ultimately, the best choice is one that is clear to your team and consistent with other `oneof` naming in your project. However, based on the information provided, `component` seems to offer the best balance of domain-specificity and clarity for this particular `oneof`. `info` is acceptable but less ideal than these alternatives.

</details>

[^1]: For example: https://github.com/protocolbuffers/protobuf/blob/c4ec2a8c0e00f3107a3aa6186e9a1e490b5bca8b/src/google/protobuf/struct.proto#L62-L78 although when searching more in depth I actually didn't find soooo many uses of `kind`.

tiyash-basu-frequenz replied:
okay, will change to kind.

tiyash-basu-frequenz replied:
Done. Ready for another look.

tiyash-basu-frequenz committed:
Rename category type field in ElectricalComponent

This commit renames the folowing:
- `ComponentCatgoryMetadataVariant` to
  `ElectricalComponentCategorySpecificInfo`
- `ElectricalComponent.category_type` to
  `ElectricalComponent.category_specific_info`

These renames make the entity names better reflect their purpose.

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>
